docs(api): add `addRouteMiddleware` util (#6894)
nuxt/framework: docs/content/3.api/3.utils/add-route-middleware.md

# `addRouteMiddleware`

`addRouteMiddleware()` is a helper function to dynamically add route middleware in your Nuxt application.

> Route middleware are navigation guards and they are generally stored in the `/middelware` directory of your Nuxt application. But using `addRouteMiddleware()` utility function, we can add route middleware directly where they are needed, such as Nuxt plugins.

Reviewer: Should be "By using ..." instead of "But using ..."

> Route middleware are navigation guards and they are generally stored in the `/middelware` directory of your Nuxt application. But using `addRouteMiddleware()` utility function, we can add route middleware directly where they are needed, such as Nuxt plugins.

Reviewer 2: ```suggestion
::alert{type=info}
Route middleware are navigation guards and are generally stored in the [`middleware/`](/guide/directory-structure/middleware) directory of your Nuxt application.
::
```

# `addRouteMiddleware`

`addRouteMiddleware()` is a helper function to dynamically add route middleware in your Nuxt application.

::alert{type=info}
Route middleware are navigation guards and are generally stored in the [`middleware/`](/guide/directory-structure/middleware) directory of your Nuxt application.
::

## Usage

```js
addRouteMiddleware (name: string | RouteMiddleware, middleware?: RouteMiddleware, options: AddRouteMiddlewareOptions = {})
```

`addRouteMiddleware()` takes three arguments:

- **name** `type: string | RouteMiddleware`

> `name` can either be a `string` or a function of type `RouteMiddleware`. This function provides the next route `to` as the first argument and the current route `from` as the second argument. Both `to` and `from` are Vue route objects.
> 
> Learn more about available properties found on [route objects](/api/composables/use-route).

Reviewer 3: ```suggestion
`name` can be either a string or a function of type `RouteMiddleware`. Function takes the next route `to` as the first argument and the current route `from` as the second argument, both of which are Vue route objects.

Learn more about available properties of [route objects](/api/composables/use-route).
```

- **middleware**: `type: RouteMiddleware`

> The second argument is also a function of type `RouteMiddleware` but it is optional. Same as mentioned above, this function provides `to` and `from` route objects. This argument becomes optional if the first argument in `addRouteMiddleware()` is already passed as a function.

Reviewer 3: ```suggestion
The second argument is a function of type `RouteMiddleware`. Same as above, it provides `to` and `from` route objects. It becomes optional if the first argument in `addRouteMiddleware()` is already passed as a function.
```

- **options:** `type: AddRouteMiddlewareOptions`  

> `options` lets you set the value of `global` to `true` to indicate whether the router middleware is global or not. By default the `global` is set to `false`.

Reviewer 3: ```suggestion
An optional `options` argument lets you set the value of `global` to `true` to indicate whether the router middleware is global or not (set to `false` by default).
```

## Examples

> ### Inline route middleware
> 
> Inline route middleware is anonymous because it does not have a name. `function` is passed as the first argument and therefore the second argument of `middleware` becomes redundant.
> 
> ```js [plugins/my-plugin.ts]

Reviewer 3: ```suggestion
### Anonymous Route Middleware

Anonymous route middleware does not have a name. It takes `function` as the first argument, making the second `middleware` argument redundant:

```ts [plugins/my-plugin.ts]
```

export default defineNuxtPlugin(() => {
  addRouteMiddleware((to, from) => {
    if (to.path === '/forbidden') {
      return false
    }
  })
})
```

> ### Named route middleware
> 
> Named route middleware always takes the first argument as a `string` and the second argument as a `function`.
> 
> Named route middleware defined through a plugin like this overrides any existing middleware of the same name located in the `/middleware` directory.
> 
> ```js [plugins/my-plugin.ts]

Reviewer 3: ```suggestion
### Named Route Middleware

Named route middleware takes `string` as the first argument and `function` as the second.

When defined in a plugin, it overrides any existing middleware of the same name located in the `/middleware` directory:

```ts [plugins/my-plugin.ts]
```

export default defineNuxtPlugin(() => {
  addRouteMiddleware('named-middleware', () => {
    console.log('named middleware added in Nuxt plugin')
  })
})
```

> ### Global route middleware
> 
> You can also set  an optional, third argument `{ global: true }` to indicate whether the route middleware being defined is `global`.
> 
> ```js [plugins/my-plugin.ts]
> export default defineNuxtPlugin(() => {
>   addRouteMiddleware('global-middleware', (to, from) => {
>       console.log('global middleware that runs on every route change')
>     }, 
>  { global: true })
> })

Reviewer 3: ```suggestion
### Global Route Middleware

You can set an optional, third argument `{ global: true }` to indicate whether the route middleware is global:

```ts [plugins/my-plugin.ts]
export default defineNuxtPlugin(() => {
  addRouteMiddleware('global-middleware', (to, from) => {
      console.log('global middleware that runs on every route change')
    },
    { global: true }
  )
})
```

Learn more about available properties of [route objects](/api/composables/use-route).

- **middleware**: `type: RouteMiddleware`

The second argument is a function of type `RouteMiddleware`. Same as above, it provides `to` and `from` route objects. It becomes optional if the first argument in `addRouteMiddleware()` is already passed as a function.

- **options:** `type: AddRouteMiddlewareOptions`  

An optional `options` argument lets you set the value of `global` to `true` to indicate whether the router middleware is global or not (set to `false` by default).

## Examples

### Anonymous Route Middleware

Anonymous route middleware does not have a name. It takes a function as the first argument, making the second `middleware` argument redundant:

```ts [plugins/my-plugin.ts]
export default defineNuxtPlugin(() => {
  addRouteMiddleware((to, from) => {
    if (to.path === '/forbidden') {
      return false
    }
  })
})
```

### Named Route Middleware

Named route middleware takes a string as the first argument and a function as the second.

> When defined in a plugin, it overrides any existing middleware of the same name located in the `/middleware` directory:

Reviewer 3: ```suggestion
When defined in a plugin, it overrides any existing middleware of the same name located in the `middleware/` directory:
```
